### PR TITLE
Add loopback support for BF devices

### LINKED ISSUE #171: Implement port loopback mode on Tofino
PR #157 added loopback mode for ports on Broadcom switches. We should have the same for Tofino devices.

### PR DESCRIPTION
**Pull request by Yi-Tseng**
This patch allows users to config loopback mode to a port via gNMI or chassis config

fix #171

### CODE REVIEW

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 1]
```diff
@@ -97,12 +97,20 @@ ::util::Status BFChassisManager::AddPortHelper(
         unit, port_id, config_params.autoneg()));
   }
   config->autoneg = config_params.autoneg();
+
+  if (config_params.loopback_mode() != LOOPBACK_STATE_UNKNOWN) {
+    LOG(INFO) << "Setting port " << port_id << " to loopback mode "
+      << config_params.loopback_mode() << ".";
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:103]: Double indent.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 2]
```diff
@@ -97,12 +97,20 @@ ::util::Status BFChassisManager::AddPortHelper(
...
+        unit, port_id, config_params.loopback_mode()));
+    config->loopback_mode = config_params.loopback_mode();
+  }
+
   if (config_params.admin_state() == ADMIN_STATE_ENABLED) {
     LOG(INFO) << "Enabling port " << port_id << " in node " << node_id << ".";
     RETURN_IF_ERROR(bf_pal_interface_->PortEnable(unit, port_id));
     config->admin_state = ADMIN_STATE_ENABLED;
   }
-
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:105]: Why remove the newline here?

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 3]
```diff
@@ -276,6 +289,20 @@ bool BFPalWrapper::PortIsValid(int unit, uint32 port_id) {
...
+::util::Status BFPalWrapper::PortLoopbackModeSet(int unit, uint32 port_id,
+                                                 LoopbackState loopback_mode) {
+  ASSIGN_OR_RETURN(bf_loopback_mode_e lp_mode, LoopbackModeToBf(loopback_mode));
+  auto bf_status =
+    bf_pal_port_loopback_mode_set(static_cast<bf_dev_id_t>(unit),
+                                  static_cast<bf_dev_port_t>(port_id),
+                                  lp_mode);
+  if (bf_status != BF_SUCCESS) {
+    return MAKE_ERROR(ERR_INTERNAL) << "Error when setting loopback mode on dev"
+                                       << unit << " port " << port_id;
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:301]: Fix indent.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 4]
```diff
@@ -188,6 +197,13 @@ ::util::Status BFChassisManager::UpdatePortHelper(
     config->autoneg = config_params.autoneg();
     config_changed = true;
   }
+  if (config_params.loopback_mode() != config_old.loopback_mode) {
+    RETURN_IF_ERROR(bf_pal_interface_->PortLoopbackModeSet(
+        unit, port_id, config_params.loopback_mode()));
+    config->loopback_mode.reset();
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:203]: Why reset and then re-assign?

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 5]
```diff
@@ -550,6 +575,12 @@ ::util::Status BFChassisManager::ReplayPortsConfig(uint64 node_id) {
           unit, port_id, *config.autoneg));
       config_new->autoneg = *config.autoneg;
     }
+    if (config.loopback_mode &&
+        config.loopback_mode != LOOPBACK_STATE_UNKNOWN) {
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:579]: Isn't this testing the same condition? `UNKNOWN = 0`

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 6]
```diff
@@ -276,6 +289,20 @@ bool BFPalWrapper::PortIsValid(int unit, uint32 port_id) {
   return (bf_pal_port_is_valid(unit, port_id) == BF_SUCCESS);
 }
 
+::util::Status BFPalWrapper::PortLoopbackModeSet(int unit, uint32 port_id,
+                                                 LoopbackState loopback_mode) {
+  ASSIGN_OR_RETURN(bf_loopback_mode_e lp_mode, LoopbackModeToBf(loopback_mode));
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:294]: Could we ignore the loopback mode UNKNOWN here? Then the other checks are not needed.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc (hunk 5 above)
Comment (Yi-Tseng) [reply stratum/hal/lib/barefoot/bf_chassis_manager.cc:579]: config.loopback_mode is an `absl::optional` type
the bool operator of `absl::optional` returns true if the value is present
and `!=` operator will compare the actual value in the `absl::optional` variable

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc (hunk 6 above)
Comment (Yi-Tseng) [reply stratum/hal/lib/barefoot/bf_pal_wrapper.cc:294]: It should be fine to return OK here if it is unknown, will do.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 7]
```diff
@@ -188,6 +197,13 @@ ::util::Status BFChassisManager::UpdatePortHelper(
     config->autoneg = config_params.autoneg();
     config_changed = true;
   }
+if (config_params.loopback_mode() != config_old.loopback_mode) {
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:200]: Indent by one.

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 8]
```diff
@@ -194,6 +194,19 @@ ::util::StatusOr<bf_fec_type_t> FecModeHalToBf(FecMode fec_mode,
...
+::util::StatusOr<bf_loopback_mode_e>
+LoopbackModeToBf(LoopbackState loopback_mode) {
+  switch (loopback_mode) {
+    case LOOPBACK_STATE_NONE:
+      return BF_LPBK_NONE;
+    case LOOPBACK_STATE_MAC:
+      return BF_LPBK_MAC_NEAR;
+    default:
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported loopback mode: "
+        << loopback_mode;
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:206]: Fix indent

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 9]
```diff
@@ -276,6 +289,24 @@ bool BFPalWrapper::PortIsValid(int unit, uint32 port_id) {
...
+    // Do nothing if we try to set loopback mode to default one (UNKNOWN).
+    return ::util::OkStatus();
+  }
+  ASSIGN_OR_RETURN(bf_loopback_mode_e lp_mode, LoopbackModeToBf(loopback_mode));
+  auto bf_status =
+      bf_pal_port_loopback_mode_set(static_cast<bf_dev_id_t>(unit),
+                                    static_cast<bf_dev_port_t>(port_id),
+                                    lp_mode);
+  if (bf_status != BF_SUCCESS) {
+    return MAKE_ERROR(ERR_INTERNAL) << "Error when setting loopback mode on dev"
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:304]: Missing space after "dev"

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 10]
```diff
@@ -276,6 +289,24 @@ bool BFPalWrapper::PortIsValid(int unit, uint32 port_id) {
...
+                                    << unit << " port " << port_id;
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:305]: I think we put a dot at the end of most logging messages.

File: stratum/hal/lib/barefoot/bf_pal_wrapper.cc [hunk 11]
```diff
@@ -194,6 +194,19 @@ ::util::StatusOr<bf_fec_type_t> FecModeHalToBf(FecMode fec_mode,
...
+::util::StatusOr<bf_loopback_mode_e>
+LoopbackModeToBf(LoopbackState loopback_mode) {
+  switch (loopback_mode) {
+    case LOOPBACK_STATE_NONE:
+      return BF_LPBK_NONE;
+    case LOOPBACK_STATE_MAC:
+      return BF_LPBK_MAC_NEAR;
+    default:
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported loopback mode: "
+          << loopback_mode;
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_pal_wrapper.cc:206]: Missing dot at end of log message.

**Pull request merged**